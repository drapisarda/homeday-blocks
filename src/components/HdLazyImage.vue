<template>
  <div
    v-if="background"
    ref="imageHolder"
    :class="{ isLoaded: imageLoaded }"
    class="lazy-image"
  />

  <!-- the src field is used as default value for the item image -->
  <!-- IE11 uses this value only because do not support the picture element -->
  <picture v-else>
      <source v-for="(source, media) in pictureSources"
        :key="media"
        :media="`(${media})`" :srcset="source"
      >
    <img ref="imageHolder" class="lazy-image" :class="{ isLoaded: imageLoaded }" :alt="alt">
  </picture>
</template>

<script>
export default {
  name: 'HdLazyImage',
  props: {
    src: {
      type: String,
      default: '',
    },
    srcset: {
      type: String,
      default: '',
    },
    srcSmall: {
      type: String,
      default: '',
    },
    alt: {
      type: String,
      default: null,
    },
    background: {
      type: Boolean,
      default: false,
    },
    pictureSources: {
      type: Array,
      default: null,
    },
  },
  data() {
    return {
      imageLoaded: false,
    };
  },
  watch: {
    src() {
      this.maybeLazyLoad();
    },
  },
  mounted() {
    this.maybeLazyLoad();
  },
  methods: {
    maybeLazyLoad() {
      this.setImageUrl('');

      if (!this.srcSmall) {
        this.setImageUrl(this.src);
        this.imageLoaded = true;
        return;
      }

      this.imageLoaded = false;
      this.setImageUrl(this.srcSmall);

      const image = new Image();

      image.onload = () => {
        this.imageLoaded = true;
        this.setImageUrl(this.src);
        this.setImageSrcset();
      };

      image.src = this.src;
    },
    setImageUrl(url) {
      if (!this.$refs.imageHolder) {
        return;
      }

      if (this.background) {
        this.$refs.imageHolder.style.backgroundImage = `url('${url}')`;
      } else {
        this.$refs.imageHolder.src = url;
      }
    },
    setImageSrcset() {
      if (!this.srcset) {
        return;
      }

      this.$refs.imageHolder.srcset = this.srcset;
    },
  },
};
</script>

<style lang="scss" scoped>
.lazy-image {
  width: 100%;
  filter: blur(4px);

  &.isLoaded {
    filter: blur(0);
  }
}
</style>
