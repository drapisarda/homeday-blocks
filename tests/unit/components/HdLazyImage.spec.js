import { mount } from '@vue/test-utils';
import HdLazyImage from '@/components/HdLazyImage.vue';
import ITEMS from '@/stories/mocks/GALLERY_ITEMS';

describe('HdLazyImage', () => {
  it('renders with the image as background', () => {
    const wrapper = mount(HdLazyImage, {
      src: ITEMS[0].image,
      background: true,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders with the image as picture element, sources and source set', () => {
    const wrapper = mount(HdLazyImage, {
      src: ITEMS[3].image,
      srcset: ITEMS[3].imageSrcSet,
      srcSmall: ITEMS[3].thumbnail,
      pictureSources: ITEMS[3].pictureSources,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
