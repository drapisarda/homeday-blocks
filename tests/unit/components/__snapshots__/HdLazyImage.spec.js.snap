// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdLazyImage renders with the image as background 1`] = `<picture> <img class="lazy-image" src=""></picture>`;

exports[`HdLazyImage renders with the image as picture element, sources and source set 1`] = `<picture> <img class="lazy-image" src=""></picture>`;
